### PR TITLE
V2routing

### LINKED ISSUE #11: Add slug type
Example:

```
/path/to/<slug:slug>
```

Perhaps something like this:

```
"slug": (str, re.compile(r"^[\w-]+")),
```

### LINKED ISSUE #12: Deprecate string
The `string` parameter type is somewhat confusing since we also have `int`. Often people new to Sanic will think it should be `str`.

I propose that it becomes deprecated, we add it as an alias to `str` for some period of time, and then remove it prior to the 21.12 Sanic LTS.

@sanic-org/framework thoughts?

### LINKED ISSUE #28: Dynamic route sorting
Currently it is this:

```python
        return (
            child.dynamic,
            len(child._children),
            key,
            bool(child.group and child.group.regex),
            type_ * -1,
        )
```

It should be:

```python
        return (
            child.dynamic,
            len(child._children),
            bool(child.group and child.group.regex),
            type_ * -1,
            key,
        )
  ```

... so that the variable name is the last tie breaker

### LINKED ISSUE #32: Alpha type
Because `alpha` types are no longer using regex, they are acting the same as `string`. We need to add a new constructor to be used inside the router in place of `str`.

### LINKED ISSUE #34: Add route.uri
This should be:

```python
@property
def uri(self):
    return f"/{self.path}"
```

### LINKED ISSUE #35: Some Sanic Routes fail with 404 Not Found
Some of our routes are not resolved. Return 404 Not Found. 
Last tested on sanic-routing v0.6.2 , sanic v21.3.4
Unittest to reproduce:

Line to add to /tests/test_routing.py
```python
def test_failing():

    def handler1():
        return "handler1"

    def handler2():
        return "handler2"

    router = Router()
    router.add("/v1/c", handler1, methods=["GET"])
    router.add("/v1/c/<c_id:int>", handler2, methods=["GET"])
    router.add("/v1/c/<c_id:int>/e", handler1, methods=["GET"])
    router.add("/v1/c/<c_id:int>/e/<e_id:int>", handler2, methods=["GET"])
    router.add("/v1/c/<c_id:int>/f", handler1, methods=["GET"])
    router.add("/v1/c/<c_id:int>/f/<f_id:int>", handler2, methods=["GET"])
    # Comment next line to pass the test
    router.add("/v1/c/<c_id:int>/d", handler1, methods=["GET"])
    router.add("/v1/c/<c_id:int>/d/<d_id:int>", handler2, methods=["GET"])
    router.finalize()


    _, handler, params = router.get("/v1/c", "GET")
    assert handler() == "handler1"
    assert params == {}

    _, handler, params = router.get("/v1/c/123", "GET")
    assert handler() == "handler2"
    assert params == {"c_id": 123}

    _, handler, params = router.get("/v1/c/123/e", "GET")
    assert handler() == "handler1"
    assert params == {"c_id": 123}

    _, handler, params = router.get("/v1/c/123/e/456", "GET")
    assert handler() == "handler2"
    assert params == {"c_id": 123, "e_id": 456}

    _, handler, params = router.get("/v1/c/123/f", "GET")
    assert handler() == "handler1"
    assert params == {"c_id": 123}

    _, handler, params = router.get("/v1/c/123/f/890", "GET")
    assert handler() == "handler2"
    assert params == {"c_id": 123, "f_id": 890}
```

Seems to have something to do with the auto generated src for compile.

### PR DESCRIPTION
**Pull request by ahopkins**
This is a pretty thorough overhaul of how the dynamic paths generated. The concept is largely still the same, but with this methodology we can take out some of the more hacky bits re: indentations. Also, in my testing it covers a wider array of edge cases including #35.

This also closes the loop on a few items including:
- Proper differentiation between `alpha` and `string` param types
- Adds a `slug` param type
- Deprecates `<foo:string>` in favor of `<foo:str>`
- Deprecates `<foo:number>` in favor of `<foo:float>`
- Adds a `route.uri` accessor

Fixes #35 
Fixes #32
Resolves #34
Resolves #12
Resolves #11
Closes #28

### CODE REVIEW

**Comment by ashleysommer** [comment]
Awesome! I'll check out the new implementation.

**Comment by ahopkins** [comment]
@sanic-org/framework thoughts?

I'd like to release this as a `rc`, and as a full version with 21.6

**Review by Tronic — APPROVED** [review]

**Comment by Tronic** [comment]
That review was only for the docs, not the other changes (not sure how I approving 2 reviewed files ended up approving 8 others without Github showing them to me).

File: sanic_routing/group.py [hunk 1]
```diff
@@ -68,55 +133,26 @@ def merge(self, group, overwrite: bool = False, append: bool = False):
...
+    def depth(self) -> int:
+        return len(self[0].parts)
 
     @property
-    def methods(self):
-        return frozenset(
-            [method for route in self for method in route.methods]
+    def dynamic_path(self) -> bool:
+        return any(
+            (param.label == "path") or (r"/" in param.label)
```
Comment (Tronic) [inline sanic_routing/group.py:142]: Unnecessary raw string for forward slash

File: sanic_routing/patterns.py [hunk 2]
```diff
@@ -7,22 +7,46 @@ def parse_date(d):
...
-    "string": (str, re.compile(r"^[^/]+")),
-    "alpha": (str, re.compile(r"^[A-Za-z]+")),
-    "path": (str, re.compile(r"^[^/]?.*?")),
-    "number": (float, re.compile(r"^-?(?:\d+(?:\.\d*)?|\.\d+)")),
-    "int": (int, re.compile(r"^-?\d+")),
+    "string": (str, re.compile(r"^[^/]+$")),
+    "str": (str, re.compile(r"^[^/]+$")),
+    "slug": (slug, re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")),
+    "alpha": (alpha, re.compile(r"^[A-Za-z]+$")),
+    "path": (str, re.compile(r"^[^/]?.*?$")),
```
Comment (Tronic) [inline sanic_routing/patterns.py:37]: Is the path regex correct? What's that trickery with optional initial non-slash?

File: sanic_routing/route.py [hunk 3]
```diff
@@ -84,21 +92,38 @@ def __str__(self):
...
-                self.parts,
+                self.segments,
                 self.requirements,
             )
             == (
-                other.parts,
+                other.segments,
                 other.requirements,
             )
             and (self.methods & other.methods)
```
Comment (Tronic) [inline sanic_routing/route.py:115]: If *any* methods match, there is equality. If this is intended (rather than requiring identical methods), a comment would be useful.

File: sanic_routing/route.py [hunk 4]
```diff
@@ -84,21 +92,38 @@ def __str__(self):
...
         )
 
+    def _ingest_path(self, path):
+        segments = []
+        for part in path.split(self.router.delimiter):
+            if "<" in part and ":" not in part:
+                name = part[1:-1]
+                part = f"<{name}:str>"
```
Comment (Tronic) [inline sanic_routing/route.py:123]: Tests for `<` anywhere in part, then assumes it to be the first character. I suggest using `part.startswith("<")`

File: sanic_routing/route.py [hunk 5]
```diff
@@ -108,25 +133,15 @@ def _setup_params(self):
...
-                            "string",
-                            str,
-                            self.router.regex_types["string"],
-                        )
+                    (
+                        name,
+                        label,
+                        _type,
+                        pattern,
+                    ) = self.parse_parameter_string(part[1:-1])
```
Comment (Tronic) [inline sanic_routing/route.py:141]: Horrible formatting. Could this be kept to a single line? (doesn't seem to be too far above the column limit).

File: sanic_routing/route.py [hunk 6]
```diff
@@ -152,6 +167,9 @@ def add_parameter(
             cast,
             pattern,
             label not in self.router.regex_types,
+            list(self.router.regex_types.keys()).index(label)
+            if label in self.router.regex_types
+            else 0,
```
Comment (Tronic) [inline sanic_routing/route.py:172]: What is at index 0 of that map? String type, presumably, but a comment here would be helpful.

File: sanic_routing/route.py [hunk 7]
```diff
@@ -221,6 +239,28 @@ def defined_params(self):
...
+        include param keys since they have no impact on routing.
+        """
+        return tuple(
+            [
+                f"<__dynamic__:{self._params[idx].label}>"
+                if self._params.get(idx)
+                else segment
+                for idx, segment in enumerate(self.parts)
+            ]
+        )
```
Comment (Tronic) [inline sanic_routing/route.py:255]: I think you can drop the square brackets, just passing a generator expression within tuple(). And perhaps use `in` rather than `get` in the condition?

File: sanic_routing/router.py [hunk 8]
```diff
@@ -367,51 +507,68 @@ def routes(self):
...
-                    if offset < 0:
-                        offset += 1
-                    else:
-                        offset = 0
-
+    def _optimize(self, node) -> None:
+        warn(
+            "Router AST optimization is an experimental only feature. "
+            "Results may vary from unoptimized code."
+        )
```
Comment (Tronic) [inline sanic_routing/router.py:514]: w00t. You really put some effort into this router.

File: sanic_routing/tree.py [hunk 9]
```diff
@@ -57,186 +88,230 @@ def display(self) -> None:
             child.display()
 
     def render(self) -> t.Tuple[t.List[Line], t.List[Line]]:
-        output: t.List[Line] = []
+        # output - code injected into the source as it is being
+        #    called/evalueated
```
Comment (Tronic) [inline sanic_routing/tree.py:92]: typo

File: sanic_routing/tree.py [hunk 10]
```diff
@@ -57,186 +88,230 @@ def display(self) -> None:
...
-                child.apply_offset(amt, apply_children=True)
+        return src + delayed, final
 
     def to_src(self) -> t.Tuple[t.List[Line], t.List[Line], t.List[Line]]:
-        indent = (self.level + 1) * 2 - 3 + self.offset
+        siblings = self.parent.children if self.parent else {}
+        first_sibling: t.Optional[Node] = None
+
+        if not self.first:
+            first_sibling = list(siblings.values())[0]
```
Comment (Tronic) [inline sanic_routing/tree.py:113]: I suppose this is not performance critical but maybe `next(iter(siblings.values()))` to avoid making a full list.

File: sanic_routing/tree.py [hunk 11]
```diff
@@ -57,186 +88,230 @@ def display(self) -> None:
...
+            # Maybe try and sneak an equality check in?
+            if_stmt = "if"
             len_check = (
                 f" and num == {self.level}"
-                if not self.children and not equality_check
+                if not self.children and not self.equality_check
                 else ""
             )
+
+            self.equality_check = self.equality_check or bool(len_check)
```
Comment (Tronic) [inline sanic_routing/tree.py:179]: `self.equality_check |= bool(len_check)` (well, a matter of style)

File: sanic_routing/tree.py [hunk 12]
```diff
@@ -258,111 +333,79 @@ def _inject_requirements(self, location, indent):
...
+            return type_ * -1
+
+        segments = tuple(map(get_type, item.parts))
+        return segments
+
+    @property
+    def depth(self):
+        if not self._children:
+            return self.level
+        return max(child.depth for child in self._children.values())
```
Comment (Tronic) [inline sanic_routing/tree.py:407]: Is this off by one, since it is directly using child's depth?

File: sanic_routing/router.py [hunk 13]
```diff
@@ -367,51 +507,68 @@ def routes(self):
...
+                nested = node.body[0].body[0]
+
+                values: t.List[t.Any] = []
+                for test in [current.test, nested.test]:
+                    if isinstance(test, ast.Compare):
+                        values.append(test)
+                    elif isinstance(test, ast.BoolOp) and isinstance(
+                        test.op, ast.And
+                    ):
+                        values.extend(test.values)
```
Comment (Tronic) [inline sanic_routing/router.py:538]: Maybe add `else` just in case something else is found so that the test is not silently dropped?

File: sanic_routing/router.py [hunk 14]
```diff
@@ -367,51 +507,68 @@ def routes(self):
...
+            #           foo1()
+            #           foo2()
             if (
-                line.src.startswith("if")
-                or line.src.startswith("elif")
-                or line.src.startswith("return")
-                or line.src.startswith("basket")
-                or line.src.startswith("try")
+                all(isinstance(child, ast.If) for child in node.body)
+                and len({child.test for child in node.body})
```
Comment (Tronic) [inline sanic_routing/router.py:556]: What is this row supposed to do?

**Review by Tronic — COMMENTED** [review]
Reviewed everything except the code generation in tree.py. Mostly stylistic issues, possibly minor bugs but no algorithmic problems that I could find so far.

I'd suggest implementing a helper class for code generation that internally keeps track of indentation to avoid manual indent management and all the clutter that comes with Line objects etc.

File: sanic_routing/tree.py [hunk 15]
```diff
@@ -57,186 +88,230 @@ def display(self) -> None:
...
-                    self._mark_parents_defer(self.parent)
-                self._inject_regex(
-                    location,
-                    return_indent,
-                    not self.parent.children_param_injected,
+            location = delayed
+
+            # Do any missing equality_check
+            if not self.equality_check:
+                conditional = "elif" if self.children else "if"
```
Comment (Tronic) [inline sanic_routing/tree.py:198]: Why is this `elif` when there are children? Maybe add a clarifying comment (or better yet, refactor if vs. elif management to a code generation class, if we always have a sequence of `if` followed only by `elif`s within a block).

File: sanic_routing/tree.py [hunk 16]
```diff
@@ -57,186 +88,230 @@ def display(self) -> None:
...
+                f"{unquote_end}",
+                indent + 1,
+            ),
+            Line("except ValueError:", indent),
+            Line("pass", indent + 1),
+            Line("else:", indent),
+        ]
+        self.base_indent += 1
+
+        location.extend(lines)
```
Comment (Tronic) [inline sanic_routing/tree.py:266]: Could use `location += lines`

File: sanic_routing/tree.py [hunk 17]
```diff
@@ -57,186 +88,230 @@ def display(self) -> None:
...
+                f"basket['__matches__'][{idx}] = "
+                f"{unquote_start}{self.param.cast.__name__}(parts[{idx}])"
+                f"{unquote_end}",
+                indent + 1,
+            ),
+            Line("except ValueError:", indent),
+            Line("pass", indent + 1),
+            Line("else:", indent),
+        ]
+        self.base_indent += 1
```
Comment (Tronic) [inline sanic_routing/tree.py:264]: I gather that the `base_indent` increment is to account for the `else:` block but `indent` is also incremented where this function is called. Is this intentional?

Comment (Tronic) [reply sanic_routing/tree.py:264]: Possible optimization: do unquoting on the whole URL, or at least only once per segment instead of separately for each test (benchmarked 250 ns per unquote call, not much but about 10 times more than str eq comparison).

File: sanic_routing/tree.py [hunk 18]
```diff
@@ -57,186 +88,230 @@ def display(self) -> None:
...
-            if k == 0:
-                location.extend(
-                    [
-                        Line(f"if num > {self.level}:", indent),
-                        Line("raise NotFound", indent + 1),
-                    ]
-                )
+    def _inject_return(self, location, indent, route_idx, group):
+        """
+        The return statement for the node if neededs
```
Comment (Tronic) [inline sanic_routing/tree.py:292]: typo

**Review by Tronic — COMMENTED** [review]
💯

File: sanic_routing/group.py (hunk 1 above)
Comment (ahopkins) [reply sanic_routing/group.py:142]: Fixed in next push.

File: sanic_routing/patterns.py (hunk 2 above)
Comment (ahopkins) [reply sanic_routing/patterns.py:37]: We have some tests for this here: https://github.com/sanic-org/sanic/blob/main/tests/test_routes.py#L582

TBH, I cannot recall the reason for making it optional. The commit is here: https://github.com/sanic-org/sanic-routing/commit/87d8adabf13a8ec6b1964f865ecf97927a615e22

It does seem like we can probably just remove it completely. Will need to do more testing though.

File: sanic_routing/route.py (hunk 3 above)
Comment (ahopkins) [reply sanic_routing/route.py:115]: Yes, this is the intention.

One could be `GET,PUT,POST` and the other is `GET`: we want them to match.

Comment in next push.

File: sanic_routing/route.py (hunk 5 above)
Comment (ahopkins) [reply sanic_routing/route.py:141]: Black is persistent. We defer to `black`.

File: sanic_routing/route.py (hunk 6 above)
Comment (ahopkins) [reply sanic_routing/route.py:172]: I think if the code is reworked slightly it becomes self-documenting:

```python
        is_regex = label not in self.router.regex_types
        priority = (
            0
            if is_regex
            else list(self.router.regex_types.keys()).index(label)
        )
        self._params[idx] = ParamInfo(
            name, raw_path, label, cast, pattern, is_regex, priority
        )
```

File: sanic_routing/route.py (hunk 7 above)
Comment (ahopkins) [reply sanic_routing/route.py:255]: In next push.

File: sanic_routing/router.py (hunk 8 above)
Comment (ahopkins) [reply sanic_routing/router.py:514]: :upside_down_face: Still plenty more to go.

File: sanic_routing/tree.py (hunk 9 above)
Comment (ahopkins) [reply sanic_routing/tree.py:92]: fixed.

File: sanic_routing/tree.py (hunk 10 above)
Comment (ahopkins) [reply sanic_routing/tree.py:113]: That makes sense.

File: sanic_routing/tree.py (hunk 12 above)
Comment (ahopkins) [reply sanic_routing/tree.py:407]: Nope. If it has children, it should get the depth of the deepest child and use that.

Here is an illustration of the intent:

```
Node (1)
Node (4)
    Node (3)
        Node (3)
    Node (4)
        Node (4)
            Node (4)
        Node (3)
```

File: sanic_routing/router.py (hunk 14 above)
Comment (ahopkins) [reply sanic_routing/router.py:556]: :thinking: I think this might be a relic from an earlier test. All `If` objects have a `test` attribute, so as it is written this would never be Falsey. IIRC, at one point I was testing for the number of unique comparisons. I think I can remove this safely.

Comment (ahopkins) [reply sanic_routing/router.py:556]: I am commenting it out for now, but will leave the code in there with a note for something to come back to next round.

**Comment by ahopkins** [comment]
> Reviewed everything except the code generation in tree.py. Mostly stylistic issues, possibly minor bugs but no algorithmic problems that I could find so far.
> 
> I'd suggest implementing a helper class for code generation that internally keeps track of indentation to avoid manual indent management and all the clutter that comes with Line objects etc.

What do you have in mind:

```python
source.add("#code at same indent")
source.add("#code indent one", 1)
source.add("#code dedent one", -1)
source.add("#code specifically at indentation=1", at=1)
```

Not a bad idea, because we could do entire blocks:

```python
source.add(f"""
try:
    something({args})
except Exception:
    ...
""")

I did have a version of it using the `ast` module directly that basically did this. But, I think it is much easier to follow using string source.

File: sanic_routing/tree.py (hunk 15 above)
Comment (ahopkins) [reply sanic_routing/tree.py:198]: added
```
# If if we have not done an equality check and there are
# children nodes, then we know there is a CHECK 1
# for the children that start at the same level, and will
# be an exclusive conditional to what is being evaluated here.
# Therefore, we can use elif
#     example:
#         if num == 7:  # CHECK 1
#             child_node_stuff
#         elif num == 6:  # CHECK 5
#             current_node_stuff
```

File: sanic_routing/tree.py (hunk 16 above)
Comment (ahopkins) [reply sanic_routing/tree.py:266]: Generally I do. I opted for `extend` since there is such a large amount of `extend/append`. I think it makes the code easier to reason about than when it is constantly shifting between operators and methods.

File: sanic_routing/tree.py (hunk 17 above)
Comment (ahopkins) [reply sanic_routing/tree.py:264]: > I gather that the `base_indent` increment is to account for the `else:` block but `indent` is also incremented where this function is called. Is this intentional?

Yes. This is one of the rare instances where we increment the `base_indent` because of its impact on subsequent nodes.



> Possible optimization: do unquoting on the whole URL, or at least only once per segment instead of separately for each test (benchmarked 250 ns per unquote call, not much but about 10 times more than str eq comparison).

:+1:  Thanks for testing that.

Comment (ahopkins) [reply sanic_routing/tree.py:264]: @Tronic 

image

Currently the code does unquoting outside of the casting. There are two possible approaches I can think of: do it early or do it late.

```
parts = tuple(map(unquote, parts))  # do once for node if needed
try:
    ...
```

or

```
try:
    ...
...
else:
    basket['__matches__'][1] = unquote(basket['__matches__'][1])
```

The second approach would follow the existing logic in doing the casting before unquoting. I am thinking maybe to go with the first approach, but it would be a change. Can you think of any reasons this would be a problem. The more I think about it, the more I think that you would want it to be unquoted, especially in the context of adding custom casting operations. (See #30)

Comment (ahopkins) [reply sanic_routing/tree.py:264]: So I was thinking that we could do it once upfront for each test. But, that will not work since some routes might be quoted, and some not. Therefore, I think the best place for this to reduce unnecessary options would be to apply inside the `else` block.

Comment (ahopkins) [reply sanic_routing/tree.py:264]: image

File: sanic_routing/router.py [hunk 19]
```diff
@@ -536,6 +536,8 @@ def _optimize(self, node) -> None:
                         test.op, ast.And
                     ):
                         values.extend(test.values)
+                    else:
+                        ...
```
Comment (Tronic) [inline sanic_routing/router.py]: What is this?

**Review by Tronic — CHANGES_REQUESTED** [review]
Other than the odd ellipsis, looks good.

**Comment by Tronic** [comment]
Re code generation: maybe take that a step forward and use *actual* inline Python code instead of string literals. Code that is syntactically valid can be loaded even if the variables don't exist, provided that the particular path never gets executed. Annotations or some other hacks could be used to inject parts into it. Well, you get the idea but this is becoming some true black magic by now.

```python
# This is router code
if need_foo:
    with codegen:
        # This is generated code
        if route == "foo":
            return handler_foo

```

**Comment by ahopkins** [comment]
Strings allow interpolation. If we prefer a code based option, I'd be more favorable towards AST. An option worth keeping on the table.
